cross_list_courses/tests/jasmine/controllers/ListControllerSpec.js

```diff
@@ -0,0 +1,310 @@
+
+            expect(scope.operationInProgress).toBe('remove');
+            expect(scope.message).toBe(null);
+
+            deleteCrosslistingDeferred.reject('failure!');
+            scope.$digest();
+
+            expect(scope.operationInProgress).toBeNull();
+            expect(scope.dtInstance.reloadData).toHaveBeenCalled();
+            expect(scope.handleAjaxErrorResponse).toHaveBeenCalled();
+            expect(scope.message.alertType).toBe('danger');
+        });
+
+        it('shows success message and cleans up on success', function() {
+            scope.removeCrosslisting(1, 1, 2);
+
+            expect(scope.operationInProgress).toBe('remove');
+            expect(scope.message).toBe(null);
+
+            deleteCrosslistingDeferred.resolve('success!');
+            scope.$digest();
+
+            expect(scope.operationInProgress).toBeNull();
+            expect(scope.dtInstance.reloadData).toHaveBeenCalled();
+            expect(scope.handleAjaxErrorResponse).not.toHaveBeenCalled();
+            expect(scope.message.alertType).toBe('success');
+        });
+
+    });
+
+    describe('submitAddCrosslisting', function() {
+        var primary = '124',
+            secondary = '456',
+            postNewCrosslistingDeferred = null;
+
+        beforeEach(function () {
+            postNewCrosslistingDeferred = $q.defer();
+            scope.dtInstance = jasmine.createSpyObj('dtInstance', ['reloadData']);
+            spyOn(scope, 'postNewCrosslisting').and
+                .returnValue(postNewCrosslistingDeferred.promise);
+            spyOn(scope, 'handleAjaxErrorResponse');
+            scope.rawFormInput.primary = primary;
+            scope.rawFormInput.secondary = secondary;
+
+            scope.submitAddCrosslisting();
+
+            expect(scope.operationInProgress).toBe('add');
+            expect(scope.message).toBe(null);
+        });
+
+        it('should make sure the postNewCrosslisting is called with the correct values', function(){
+            expect(scope.postNewCrosslisting)
+                .toHaveBeenCalledWith(primary, secondary);
+        });
+
+        it('shows correct scope.message on failure', function(){
+            postNewCrosslistingDeferred.resolve('success!');
+            scope.$digest();
+
+            expect(scope.operationInProgress).toBeNull();
+            expect(scope.dtInstance.reloadData).toHaveBeenCalled();
+            expect(scope.handleAjaxErrorResponse).not.toHaveBeenCalled();
+            expect(scope.message.alertType).toBe('success');
+        });
+
+        it('shows correct scope.message on failure', function(){
+            postNewCrosslistingDeferred.reject('failure!');
+            scope.$digest();
+
+            expect(scope.operationInProgress).toBeNull();
+            expect(scope.dtInstance.reloadData).not.toHaveBeenCalled();
+            expect(scope.handleAjaxErrorResponse).toHaveBeenCalled();
+            expect(scope.message.alertType).toBe('danger');
+        });
+
```

Comment:
Would it make sense to add a  test case for the special error message handling for the duplicate? (most messages come from backend, but the one that is handled in postFailed for the duplicate)
https://github.com/Harvard-University-iCommons/canvas_account_admin_tools/blob/develop/cross_list_courses/static/cross_list_courses/js/controllers/ListController.js#L140